ext/pbr_ext/read.cpp

```diff
@@ -0,0 +1,172 @@
+// VALUE val - the field value on obj being written
+
+// these macros are unhygienic, but that's ok since they are
+// local to this file.
+#define DEF_RV(type)  VALUE rv_##type(ss_t& ss, Msg& msg, Fld& fld, vector<LazyField> *lazy_fields)
+#define INFLATE(val)  RTEST(fld.inflate) ? rb_funcall(fld.inflate, ID_CALL, 1, (val)) : (val)
+
+DEF_RV(INT32)    { return (INT2NUM          (r_varint32(ss)));  }
+DEF_RV(UINT32)   { return (UINT2NUM         (r_varint32(ss)));  }
+DEF_RV(INT64)    { return (LL2NUM           (r_varint64(ss)));  }
+DEF_RV(UINT64)   { return (ULL2NUM          (r_varint64(ss)));  }
+DEF_RV(SINT32)   { return (INT2NUM (zz_dec32(r_varint32(ss)))); }
+DEF_RV(SINT64)   { return (LL2NUM  (zz_dec64(r_varint64(ss)))); }
+DEF_RV(SFIXED32) { return (INT2NUM          (r_int32   (ss)));  }
+DEF_RV(FIXED32)  { return (UINT2NUM         (r_int32   (ss)));  }
+DEF_RV(SFIXED64) { return (LL2NUM           (r_int64   (ss)));  }
+DEF_RV(FIXED64)  { return (ULL2NUM          (r_int64   (ss)));  }
+
+DEF_RV(ENUM) {
+  VALUE val = INT2NUM(r_varint32(ss)); 
+  if (msg.model->validate_on_read)
+    validate_enum(msg, fld, val);
+  return val;
+}
+
+DEF_RV(FLOAT) {
+  uint32_t v = r_int32(ss);
+  return (DBL2NUM((double)REINTERPRET(float, v)));
+}
+
+DEF_RV(DOUBLE) {
+  uint64_t v = r_int64(ss);
+  return (DBL2NUM(REINTERPRET(double, v)));
+}
+
+DEF_RV(BOOL) {
+  uint32_t v = r_varint32(ss);
+  return (v == 0 ? Qfalse : Qtrue);
+}
+
+DEF_RV(STRING) {
+  int32_t len = r_varint32(ss);
+  VALUE rstr = rb_str_new(ss_read_chars(ss, len), len);
+  return rb_funcall(rstr, FORCE_ID_ENCODING, 1, UTF_8_ENCODING);
+}
+
+DEF_RV(BYTES) {
+  int32_t len = r_varint32(ss);
+  if (fld.get_lazy_type == Qnil) {
+    return rb_str_new(ss_read_chars(ss, len), len);
+  } else {
+    LazyField lf;
+    lf.ss = ss_substream(ss, len);
+    lf.fld = &fld;
+    lazy_fields->push_back(lf);
+    return Qnil;
+  }
+}
+
+DEF_RV(MESSAGE) {
+  int32_t len = r_varint32(ss);
+  ss_t tmp_ss = ss_substream(ss, len);
+  Msg& embedded_msg = *fld.embedded_msg;
+  return embedded_msg.read(tmp_ss, embedded_msg);
+}
+
+read_val_func get_val_reader(fld_t fld_type) {
+  switch (fld_type) {
+    TYPE_MAP(rv);
+    default:
+      rb_raise(rb_eStandardError, "I don\'t know how to read field type %d", fld_type);
+  }
+}
+
+VALUE read_obj(ss_t& ss, Msg& msg) {
+  VALUE obj = rb_funcall(msg.target, ID_CTOR, 0);
+  
+  for (Fld& fld : msg.flds_to_enumerate)
+    if (fld.label == LABEL_REPEATED)
+      set_value(msg, fld, obj, rb_ary_new());
+
+  int num_required_fields_read = 0;
+
+  vector<LazyField> lazy_fields;
+
+  while (ss_more(ss)) {
+    uint32_t h = r_varint32(ss);
+    fld_num_t fld_num = h >> 3;
+    Fld *fld_ptr = msg.find_fld(msg, fld_num);
+
+    if (fld_ptr == NULL) {
+      //cerr << "skipping unrecognized field " << msg.name << "." << fld_num << endl;
+      wire_t wire_type = h & 7;
+      skip(ss, wire_type);
+      continue;
+    }
+
+    Fld& fld = *fld_ptr;
```

Comment:
Same thing here. You're not actually doing anything unsafe, but this line of code by itself feels weird.

Keep it if you like. Otherwise, you can just use fld_ptr->label, flb_label->is_packed, etc.